✨ Support `Annotated` approach for `Relationship` attributes

bolu61 opened the pull request:
see https://github.com/fastapi/sqlmodel/issues/229#issuecomment-2132852311

I'm not really sure how Pydantic does it, but here's a potential workaround to the problem.

bolu61 committed:
support PEP 593 for relationships

bolu61 committed:
Fix formatting

bolu61 committed:
Use Optional instead of `|`

bolu61 committed:
Fix type annotation

bolu61 committed:
Use Any instead of type[Any]

svlandeg committed:
Merge branch 'main' into main

bolu61 committed:
Merge branch 'main' into main

bolu61 committed:
Merge branch 'main' into main

YuriiMotov reviewed (commented):
@bolu61, thanks for working on this!

So, this PR fixes the following code example:
```py
class Foo(SQLModel, table=True):
    id: Annotated[int | None, Field(primary_key=True)] = None
    uuid: Annotated[UUID, Field(unique=True)]

class Bar(SQLModel, table=True):
    id: Annotated[int | None, Field(primary_key=True)] = None
    uuid: Annotated[UUID, Field(unique=True)]
    foo_id: Annotated[int, Field(foreign_key="foo.id")]
    foo: Annotated[Foo, Relationship()]
```
For now last line (with `Relationship()`) doesn't work (`ValueError: <class '__main__.Foo'> has no matching SQLAlchemy type`) and this PR fixes that.

I think this is not the same that was initially proposed in #229. The idea there was to support passing multiple annotations like:
```py
class Hero(SQLModel, table=True):
    id: Annotated[Optional[int], Field(examples=....), Column(primary_key=True)] = None
```
Where `Field` can be `pydantic.Field`.
So, I suggest we unmark this PR as the one that resolves that issue.

As for changes introduced by this PR, I think it goes in line with modern recommendations to use `Annotated` approach instead parameterizing fields via default value.
I think we should test this approach more and add some automatic tests. Then it will have all chances to be accepted.

YuriiMotov commented:
> I'm not too familiar with this codebase. Can you point me to where I can write those tests?

You can create an annotated version of the code example from "Relationships" section of docs:
* Go to https://github.com/fastapi/sqlmodel/tree/main/docs_src/tutorial/relationship_attributes/define_relationship_attributes
* Create a copy of `tutorial001_py310.py`, call it `tutorial001_an_py310.py`
* Edit code to follow the Annotated approach
* Go to https://github.com/fastapi/sqlmodel/tree/main/tests/test_tutorial/test_relationship_attributes/test_define_relationship_attributes
* Create a copy of `test_tutorial001_py310.py`, call it `test_tutorial001_an_py310.py`
* Update code to import app from `tutorial001_an_py310.py`
* Push changes and check that new version of code example appeared in "Other versions and variants"

bolu61 committed:
Merge branch 'fastapi:main' into main

bolu61 committed:
Fix handling of Annotated types in relationship annotations

bolu61 committed:
📝 Add tutorial and tests for defining relationship attributes in SQLModel

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

bolu61 commented:
The test is failing with:
```
FAILED tests/test_tutorial/test_relationship_attributes/test_define_relationship_attributes/test_tutorial001_an_py310.py::test_tutorial - sqlalchemy.exc.ArgumentError: Mapper Mapper[Team(team)] could not assemble any primary key columns for mapped table 'team'
```
I'm not exactly sure how to approach this.

bolu61 committed:
Fix type hinting for relationship annotations in SQLModelMetaclass

svlandeg committed:
Merge branch 'main' into main_bolu61

pre-commit-ci-lite[bot] committed:
🎨 Auto format

YuriiMotov committed:
Simplify test

YuriiMotov reviewed (changes requested):
This needs some more work and testing..

@bolu61, would you like to continue working on this?

bolu61 committed:
Merge branch 'main' into bolu61/main

bolu61 committed:
Address review feedback for Annotated relationships

- Detect annotated relationships without a default value by iterating
  over the union of class_dict and original_annotations.
- Use elif and unwrap inner Mapped[T] when handling Annotated[Mapped[T], ...]
  to avoid double-wrapping in Mapped.
- Split the relationship/pydantic partition into three single-purpose
  loops with comments for clarity.
- Add tests covering Annotated relationships with default value, without
  default value, and with Annotated[Mapped[T], ...].

bolu61 commented on sqlmodel/main.py:
```diff
@@ -551,7 +563,11 @@ def __new__(
+            r = get_annotated_relationshipinfo(a)
+            if r is not None:
+                relationships[k] = r
+            elif isinstance(v, RelationshipInfo):
```
I'm moving the relationshipinfo logic into a dedicated loop so it handles the cases clearly.
